platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java

```diff
@@ -16,53 +16,51 @@
 
 package org.gradle.api.internal.tasks.compile;
 
-import org.gradle.api.JavaVersion;
 import org.gradle.api.tasks.compile.CompileOptions;
 import org.gradle.internal.Factory;
 import org.gradle.internal.jvm.Jvm;
 import org.gradle.jvm.toolchain.JavaInstallationMetadata;
 import org.gradle.jvm.toolchain.internal.JavaExecutableUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import javax.annotation.Nullable;
 import java.io.File;
 
 public abstract class AbstractJavaCompileSpecFactory<T extends JavaCompileSpec> implements Factory<T> {
-    private final CompileOptions compileOptions;
+    private final static Logger LOGGER = LoggerFactory.getLogger(AbstractJavaCompileSpecFactory.class);
 
+    private final CompileOptions compileOptions;
     private final JavaInstallationMetadata toolchain;
 
-    public AbstractJavaCompileSpecFactory(CompileOptions compileOptions, @Nullable JavaInstallationMetadata toolchain) {
+    public AbstractJavaCompileSpecFactory(CompileOptions compileOptions, JavaInstallationMetadata toolchain) {
         this.compileOptions = compileOptions;
         this.toolchain = toolchain;
     }
 
     @Override
     public T create() {
```

Comment:
I think we should twist the method around to make forking the default path and in-process execution an exception, I think something like this:
```
    @Override
    public T create() {
        File toolchainJavaHome = toolchain.getInstallationPath().getAsFile();
        if (!toolchain.getLanguageVersion().canCompileOrRun(8)) {
            LOGGER.info("Falling back to command line compiler: requested toolchain is below JDK 1.7");
            return getCommandLineSpec(Jvm.forHome(toolchainJavaHome).getJavacExecutable());
        }

        if (compileOptions.isFork()) {
            File forkJavaHome = compileOptions.getForkOptions().getJavaHome();
            if (forkJavaHome != null) {
                LOGGER.info("Falling back to command line compiler: forking with Java Home set");
                return getCommandLineSpec(Jvm.forHome(forkJavaHome).getJavacExecutable());
            }

            String forkExecutable = compileOptions.getForkOptions().getExecutable();
            if (forkExecutable != null) {
                LOGGER.info("Falling back to command line compiler: forking with javac executable set");
                return getCommandLineSpec(JavaExecutableUtils.resolveExecutable(forkExecutable));
            }
        }

        if (toolchain.isCurrentJvm() && JdkJavaCompiler.canBeUsed()) {
            // Please keep it in mind, that when using TestKit with debug enabled (i.e. in embedded mode), this line won't be reached after Java 16 (JEP 396)
            // If you need this to be executed, add the necessary configs from JPMSConfiguration to the test runner executing Gradle
            LOGGER.info("Using in-process compilation");
            return getDefaultSpec();
        }

        LOGGER.info("Running compilation in compiler daemon");
        return getForkingSpec(toolchainJavaHome, toolchain.getLanguageVersion().asInt());
    }
```

Reply:
Separately, we could rename `getDefaultSpec` to `getInProcessSpec` or similar.

Reply:
I agree that it seems like `getForkingSpec` became a go-to fallback, and it should work in all circumstances.

Reply:
> I think we should twist the method around to make forking the default path and in-process execution an exception, I think something like this...

Agreed, and used your version.

> Separately, we could rename getDefaultSpec to getInProcessSpec or similar.

Done